extensions/amp-a4a/0.1/head-validation.js

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {map} from '../../../src/utils/object';
+import {parseExtensionUrl} from '../../../src/service/extension-location';
+import {removeElement} from '../../../src/dom';
+
+/**
+ * @typedef {{
+ *    extensions: !Array<{extensionId: (string|undefined), extensionVersion: (string|undefined)}>,
+ *    head: !Element
+ * }}
+ */
+export let ValidatedHeadDef;
+
+const FONT_ALLOWLIST = map({
+  'https://cdn.materialdesignicons.com': true,
+  'https://cloud.typography.com': true,
+  'https://fast.fonts.net': true,
+  'https://fonts.googleapis.com': true,
+  'https://maxcdn.bootstrapcdn.com': true,
+  'https://p.typekit.net': true,
+  'https://pro.fontawesome.com': true,
+  'https://use.fontawesome.com': true,
+  'https://use.typekit.net': true,
+});
+
+const EXTENSION_ALLOWLIST = map({
+  'amp-accordion': true,
+  'amp-ad-exit': true,
+  'amp-analytics': true,
+  'amp-anim': true,
+  'amp-animation': true,
+  'amp-audio': true,
+  'amp-bind': true,
+  'amp-carousel': true,
+  'amp-fit-text': true,
+  'amp-font': true,
+  'amp-form': true,
+  'amp-img': true,
+  'amp-layout': true,
+  'amp-lightbox': true,
+  'amp-mraid': true,
+  'amp-mustache': true,
+  'amp-pixel': true,
+  'amp-position-observer': true,
+  'amp-selector': true,
+  'amp-social-share': true,
+  'amp-video': true,
+});
+
+const EXTENSION_URL_PREFIX = /^https:\/\/cdn\.ampproject.org\/v0\//;
+
+/**
+ * Sanitizes AMPHTML Ad head element and extracts extensions to be installed.
+ * @param {!Window} win
+ * @param {!Element} adElement
+ * @param {?Element} head
+ * @return {?ValidatedHeadDef}
+ */
+export function validateHead(win, adElement, head) {
+  if (!head || !head.firstChild) {
+    return null;
+  }
+
+  const extensionService = Services.extensionsFor(win);
+  const urlService = Services.urlForDoc(adElement);
+  const extensions = [];
+  const fonts = [];
+  const images = [];
+
+  let element = head.firstChild;
+  while (element) {
+    // Store next element here as the following code will remove
+    // certain elements from the detached DOM.
+    const nextElement = element.nextElementSibling;
+
+    switch (element.tagName) {
```

Comment:
is tagName guaranteed to be upperCase across browsers?

Reply:
asked because I saw many usages of `tagName.toUpperCase()` in our code base.

Reply:
makes sense, you think I should add it just to be safe?

Reply:
I would do so, just to be safer. (browsers can be weird)